### PR TITLE
V3 recovery

### PR DESCRIPTION
**Pull request by ashelly**
Same code as before, except no "on-demand braking".
Tested on 3.

Features:

Homes on rising edge of sensor, not falling (like v2.1)
limits distance to find falling edge to 1mm.
Reports position of rising and falling edge, for sensor failure detection
needs "has_sensor_reporting" flag in motion/control.json.
Add Separate alarm messages for:
Hard Limit, Soft Limit, Abort, Probe Fail, Homing Fail, Estop Pressed.
$Z command to clear all queued motions and set to IDLE.
removed stale/unused code.

### CODE REVIEW

File: system.c [hunk 1]
```diff
@@ -137,10 +140,17 @@ uint8_t system_execute_line(char *line)
       if ( line[++char_counter] != 0 ) { return(STATUS_INVALID_STATEMENT); }
       IO_RESET_PORT |= IO_RESET_MASK;  //reset IO.  Will re-enable in loop
       break;
-      
-      
+  case 'Z': //zero the current buffers
```
Comment (kborer) [inline system.c]: spacing seems different from the other case statements

File: limits.c [hunk 2]
```diff
@@ -125,78 +124,86 @@ void limits_go_home(uint8_t cycle_mask)
...
+  homing_rate = min_seek_rate * sqrt(n_active_axis); //Adjust so individual axes all move at homing rate.
 
   plan_reset(); // Reset planner buffer to zero planner current position and to clear previous motions.
 
   do {
     // Set target location and rate for active axes.
     // and reset homing axis locks based on cycle mask.  
-    uint8_t axislock = 0; 
-    uint8_t n_active_axis = 0;
+    float travel = approach?max_travel:settings.homing_pulloff; //limit travel away from the switch to the pulloff distance.
```
Comment (kborer) [inline limits.c]:
``` c
    // Set target location and rate for active axes and reset homing axis
    // locks based on cycle mask. Limit travel away from switch to pulloff.
    float travel = approach ? max_travel : settings.homing_pulloff;
```

File: limits.c [hunk 3]
```diff
@@ -125,78 +124,86 @@ void limits_go_home(uint8_t cycle_mask)
...
-    limit_approach = approach;  //limit_approach bits is high if approaching limit switch 
-
     // Perform homing cycle. Planner buffer should be empty, as required to initiate the homing cycle.
-    plan_buffer_line(target, homing_rate, false, LINENUMBER_EMPTY_BLOCK); // Bypass mc_line(). Directly plan homing motion.
+    plan_buffer_line(target, homing_rate, false, LINENUMBER_EMPTY_BLOCK);  // Bypass mc_line(). Directly plan homing motion.
 
 	 // axislock bit is high if axis is homing, so we only enable checking on moving axes.
     limits_enable(axislock,~approach);  //expect 0 on approach (stop when 1). vice versa for pulloff
-    limits.homenext =0;
+    limits.ishoming =1;
```
Comment (kborer) [inline limits.c]:
``` c
limits.ishoming = 1;
```

File: config.h [hunk 4]
```diff
@@ -80,7 +80,6 @@
 #define HOMING_CYCLE_0 (1<<Z_AXIS)                // REQUIRED: First move
```
Comment (mdicicco) [inline config.h]: Are these homing cycles used at all? Something looks wrong if we do.  Also, the "NOTE:" at the end of the comment is wrong.

File: limits.c [hunk 5]
```diff
@@ -125,78 +124,86 @@ void limits_go_home(uint8_t cycle_mask)
...
-    uint8_t axislock = 0; 
-    uint8_t n_active_axis = 0;
+    float travel = approach?max_travel:settings.homing_pulloff; //limit travel away from the switch to the pulloff distance.
     for (idx=0; idx<N_AXIS; idx++) {
       if (bit_istrue(cycle_mask,bit(idx))) {
-        n_active_axis++;
-        axislock |= (1<<(X_STEP_BIT+idx)); //assumes axes are in bit order.
-        if ((flipped&(1<<idx))^approach) { target[idx] = -max_travel; }
-        else {                             target[idx] = max_travel; }
+        if ((flipped&(1<<idx))^approach) { target[idx] = -travel; }
```
Comment (mdicicco) [inline limits.c]: This formatting is gross and non-standard.

File: limits.c [hunk 6]
```diff
@@ -125,78 +124,86 @@ void limits_go_home(uint8_t cycle_mask)
...
+      protocol_execute_runtime();
+    }
+  
     delay_ms(settings.homing_debounce_delay); // Delay to allow transient dynamics to dissipate.
 
+
     // Reverse direction and reset homing rate for locate cycle(s).
-    homing_rate = settings.homing_feed_rate;
     approach = ~approach; //toggle all bits
+    homing_rate = settings.homing_feed_rate * sqrt(n_active_axis); ;
```
Comment (mdicicco) [inline limits.c]: Two ;;'s

File: config.h (hunk 4 above)
Comment (ashelly) [inline config.h]: They are not used by the python client. However if you issued a raw $H on the serial interface they would be used. 
I will change to X,Y,Z+C, instead.

**Pull request merged**